prusti-viper/src/encoder/encoder.rs

```diff
@@ -1200,15 +1200,20 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
             }
             tymap_stack.push(tymap);
         }
-        let cleanup = || {
-            // FIXME: this is a hack to support generics. See issue #187.
-            let mut tymap_stack = self.typaram_repl.borrow_mut();
-            tymap_stack.pop();
-        };
+        // FIXME: this is a hack to support generics. See issue #187.
+        struct CleanupTyMapStack<'a, 'tcx> {
+            tymap_stack: &'a std::cell::RefCell<Vec<HashMap<ty::Ty<'tcx>, ty::Ty<'tcx>>>>,
+        }
+        impl<'a, 'tcx> Drop for CleanupTyMapStack<'a, 'tcx> {
+            fn drop(&mut self) {
+                self.tymap_stack.borrow_mut().pop();
+            }
+        }
+        let _cleanup_token = CleanupTyMapStack { tymap_stack: &self.typaram_repl };
```

Comment:
Is it guaranteed by Rust that `Drop` will only run after everything else in this function runs? I.e. could the `Drop` run right here, since `_cleanup_token` is not used anywhere else?

Reply:
yes. see e.g. [this thread](https://users.rust-lang.org/t/do-variables-shorten-the-scope-lifetime/5754). rustc uses this mechanism internally for e.g. profiling or span enter/leave; `git grep _[a-z]` on the rustc source yields quite some results.

Reply:
I had the same feeling, but it seems fine: https://rust-unofficial.github.io/patterns/idioms/dtor-finally.html

Reply:
[the rust reference](https://doc.rust-lang.org/reference/destructors.html#drop-scopes) seems to be the most official thing i found on the subject; which also only talks about blocks and scopes, not other analyses (e.g. NLL).

Reply:
I would add `#[must_use]` to the struct; it doesn't hurt.